File: week 2/ai-software-engineering/app/api/predict/route.py
from src.predict import DiseasePredictor
from lib.utils import validate_input_data, generate_risk_recommendations
import json

predictor = DiseasePredictor()

async def POST(request):
    try:
        data = await request.json()
        
        validation = validate_input_data(data)
        if 'error' in validation:
            return Response(
                json.dumps({'error': validation['error']}),
                status=400,
                headers={'Content-Type': 'application/json'}
            )
        
        prediction_result = predictor.predict(validation['data'])
        
        if 'error' in prediction_result:
            return Response(
                json.dumps({'error': prediction_result['error']}),
                status=500,
                headers={'Content-Type': 'application/json'}
            )
        
        recommendations = generate_risk_recommendations(
            prediction_result['risk_level'],
            prediction_result['confidence']
        )
        
        final_result = {
            **prediction_result,
            'recommendations': recommendations
        }
        
        return Response(
            json.dumps(final_result),
            status=200,
            headers={'Content-Type': 'application/json'}
        )
        
    except Exception as e:
        return Response(
            json.dumps({'error': f'Internal server error: {str(e)}'}),
            status=500,
            headers={'Content-Type': 'application/json'}
        )

File: week 2/ai-software-engineering/app/api/predict/route.ts
import { NextResponse } from 'next/server';

export async function POST(request: Request) {
  try {
    const inputData = await request.json();
    console.log('📊 Received prediction request:', inputData);

    // Calculate risk score
    const riskScore = 
      (inputData.temperature - 36.5) * 0.2 +
      (inputData.humidity - 50) * 0.1 +
      inputData.rainfall * 0.05 +
      (inputData.population_density - 300) * 0.001 -
      inputData.sanitation_score * 0.3 -
      inputData.vaccination_rate * 0.02 +
      inputData.previous_cases * 0.15 -
      inputData.healthcare_access * 0.2;

    const riskProbability = Math.min(0.95, Math.max(0.05, 1 / (1 + Math.exp(-riskScore * 0.1))));
    
    let riskLevel = 'Low Risk';
    if (riskProbability > 0.7) riskLevel = 'High Risk';
    else if (riskProbability > 0.4) riskLevel = 'Medium Risk';

    const recommendations = {
      'High Risk': {
        alert_level: 'HIGH',
        recommendations: [
          'Activate emergency response protocol',
          'Increase hospital preparedness and bed capacity',
          'Deploy rapid testing teams to affected areas',
          'Launch public health emergency communication'
        ],
        immediate_actions: [
          'Conduct door-to-door health screening',
          'Isolate and test suspected cases immediately',
          'Enhance vector control and sanitation measures'
        ]
      },
      'Medium Risk': {
        alert_level: 'MEDIUM',
        recommendations: [
          'Enhance disease surveillance systems',
          'Increase sanitation in public spaces',
          'Conduct public health education campaigns',
          'Review and update contingency plans'
        ],
        immediate_actions: [
          'Increase testing in suspected areas',
          'Strengthen primary healthcare services',
          'Monitor environmental conditions daily'
        ]
      },
      'Low Risk': {
        alert_level: 'LOW',
        recommendations: [
          'Maintain routine health monitoring',
          'Continue environmental surveillance',
          'Regular public health education',
          'Update disease prevention protocols'
        ],
        immediate_actions: [
          'Continue standard surveillance',
          'Monitor weather and outbreak reports',
          'Maintain routine health services'
        ]
      }
    };

    const result = {
      risk_level: riskLevel,
      confidence: 0.85 + (Math.random() * 0.1),
      risk_probability: riskProbability,
      recommendations: recommendations[riskLevel as keyof typeof recommendations]
    };

    console.log('✅ Prediction result:', result);
    return NextResponse.json(result);

  } catch (error) {
    console.error('❌ Prediction error:', error);
    return NextResponse.json(
      { error: 'Failed to process prediction' },
      { status: 500 }
    );
  }
}

export async function GET() {
  return NextResponse.json({ 
    message: 'Disease Prediction API is working!',
    status: 'OK',
    timestamp: new Date().toISOString()
  });
}